src/worldfactory.cpp

```diff
@@ -169,6 +169,26 @@ WORLD *worldfactory::make_new_world( bool show_prompt, const std::string &world_
     return add_world( std::move( retworld ) );
 }
 
+static cata::optional<std::string> prompt_world_name( const std::string &title,
+        const std::string &cur_worldname )
+{
+    string_input_popup popup;
+    popup.max_length( max_worldname_len ).title( title ).text( cur_worldname );
+    std::vector<input_event> evs =
+        inp_mngr.get_input_for_action( "PICK_RANDOM_WORLDNAME", "WORLDGEN_CONFIRM_DIALOG" );
```

Comment:
I think this will return input events not usable in the current context (such as keycode events in symbol mode or gamepad key when gamepad is not connected), so the front element is not necessarily a valid hotkey. And even if it is a valid hotkey, simply using `get_first_input` will discard modifier keys inside keycode mode.

I suggest allowing `string_input_popup` to handle additional actions similar to how it is done in `uilist`.

Reply:
Cheers. I changed the callbacks so that they can be added by raw input type or by action name. Popup text can also be modified within the callback using `string_input_popup::text()`.